### ISSUE TITLE
Compilation fails on updated esphome version

### ISSUE THREAD

**Issue by Basti-Fantasti** [evgeni/esphome-somfy#6]
Hi there,

I've been using the shelly2.5 with your config now for quite a while.
Thanks again - it works really great 👍 

Now on one of the previous updates of ESPHome, the compilation does not work anymore.
When I try to send an updated config to the shellys I get this parsing error:

`
    
    Failed config
    
    switch.template: [source /config/esphome/.shelly25.yaml:136]
      
      Either optimistic mode must be enabled, or turn_on_action or turn_off_action must be set, to handle the switch being set.
      platform: template
      name: Move UP
      internal: True
      lambda: |-
        if (id(switch2).state && !id(block_control).state) {
          return true;
        } else {
          return false;
        }
      on_turn_on: 
        then: 
          cover.open: rolladen
      on_turn_off: 
        then: 
          cover.stop: rolladen
    switch.template: [source /config/esphome/.shelly25.yaml:152]
      
      Either optimistic mode must be enabled, or turn_on_action or turn_off_action must be set, to handle the switch being set.
      platform: template
      name: Move DOWN
      internal: True
      lambda: |-
        if (id(switch1).state && !id(block_control).state) {
          return true;
        } else {
          return false;
        }
      on_turn_on: 
        then: 
          cover.close: rolladen
      on_turn_off: 
        then: 
          cover.stop: rolladen
    
`

I've looked around and found that this issue should be related to the **on_turn_on** and **on_turn_off** events.

I found another implementation for other switches where the on_turn_* events have been replaced with turn_on_action and turn_off_action. See here: [Link](https://github.com/esphome/issues/issues/548)

But I'm not sure if this command can be replaced just like this in your template here...

Any help is greatly appreciated

Best regards
Bastian

**Comment by evgeni** [comment]
I think you should be able to replace

```yaml
  on_turn_on: 
    then: 
      cover.open: rolladen
  on_turn_off: 
    then: 
      cover.stop: rolladen
```

with

```yaml
    turn_on_action:
      - cover.open: rolladen
    turn_off_action:
      - cover.stop: rolladen
```

But I don't have my esphome environment in front of me to test that out.

**Comment by Basti-Fantasti** [comment]
yes this worked 👍 

Thanks!

**Comment by evgeni** [comment]
Reopening, as I will need to fix it in my code for good and this is a perfect reminder ;)

**Comment by Basti-Fantasti** [comment]
it's good that you've opened the issue again.
Because this is only half the way 😄 
Now compilation is possible, and I can open and close the covers using the home assistant app but the physical buttons are not working reliably anymore.

I think this goes together with the blocking state or something. 
This morning the covers were closed. But I was not able to open them using the switches on the wall.
So I had to hit the **close** buttons again for several seconds - then undo them
and then it was working. 

So I assume that the progress or the actual position on where the shutters currently are is messed up.

It's only a question of time when my wife will find out that she can not manually open and close the covers anymore 😏

**Comment by evgeni** [comment]
Some notes:

The change that seems to "break" it is [Template Select/Number/Switch](https://esphome.io/changelog/2021.9.0.html#template-select-number-switch), quoting the changelog:

> The template selects, numbers and switches now have more validations on them preventing weird usage. If you do not set a lambda then you are required to make them either optimistic: true, or utilise the set_action (select/number) / turn_on_trigger (switch).

And the corresponding PR is https://github.com/esphome/esphome/pull/2230

And I can confirm your observation, app (HA) works, while physical buttons do not after the above proposed patch.

From the logs:

```
[21:52:42][D][binary_sensor:036]: 'Switch #1': Sending state ON
[21:52:42][D][switch:037]: 'Move UP': Sending state ON
[21:52:44][D][binary_sensor:036]: 'Switch #1': Sending state OFF
[21:52:44][D][switch:037]: 'Move UP': Sending state OFF
```

Which seems to imply the switch correctly triggers the state change, but no action is performed on the cover.

**Comment by evgeni** [comment]
Thanks again for the report @Basti-Fantasti!

In https://github.com/evgeni/esphome-configs/commit/83adca8fd6e84903f9813e985f57d0fccec29d3d I've added a fix for this issue - setting `optimistic: true` was correct after all. Quoting the commit message:

esphome >= 2021.9.0 requires switches to either be optimistic, or have
turn_on_action and turn_off_action defined.

the turn_*_actions are only triggered when the switch is hit via the
API [1]:

> The action that should be performed when the remote (like Home Assistant’s
> frontend) requests the switch to be turned on.

However, we set these as "internal: true" as we don't need them to be
exposed in HA at all, so we can also set "optimistic: true" and move on
with our lives.

[1] https://esphome.io/components/switch/template.html
